### ISSUE TITLE
Building 1.48 fails

### ISSUE THREAD

**Issue by yancyribbens** [rust-bitcoin/rust-bitcoin#1969]
I thought this used to work, but now when I try to build with 1.48 toolchain, I get an error:

```
cargo +1.48 build
    Updating crates.io index
error: failed to select a version for the requirement `once_cell = "=1.18.0"`
candidate versions found which didn't match: 1.17.2, 1.17.1, 1.17.0, ...
location searched: crates.io index
required by package `wasm-bindgen-backend v0.2.87`
    ... which is depended on by `wasm-bindgen-macro-support v0.2.87`
    ... which is depended on by `wasm-bindgen-macro v0.2.87`
    ... which is depended on by `wasm-bindgen v0.2.87`
    ... which is depended on by `console_error_panic_hook v0.1.7`
    ... which is depended on by `wasm-bindgen-test v0.3.37`
    ... which is depended on by `bitcoin_hashes v0.12.0 (/usr/home/yancy/Git/rust-bitcoin/hashes)`
    ... which is depended on by `bitcoin v0.30.0 (/usr/home/yancy/Git/rust-bitcoin/bitcoin)`
    ... which is depended on by `bitcoin-fuzz v0.0.1 (/usr/home/yancy/Git/rust-bitcoin/fuzz)`
```

**Comment by tcharding** [comment]
I think we should remove the versions from the README and just point devs towards the `contrib/test.sh` file, its one less thing to maintain and one is never 100% sure that the README is correct anyway, the test file is the single source of truth.

**Comment by sanket1729** [comment]
Another issue with our current CI setup is that we are not actually testing against the latest dependencies in the CI. A new user on stable would expect `cargo update && cargo test` to succeed, but we are actually not testing that in the CI setup.

**Comment by apoelstra** [comment]
I guess we should add that as a separate job (probably with `--all-features`) so we'll notice when stuff breaks.

**Comment by yancyribbens** [comment]
`cp Cargo-recent.lock ./Cargo.lock` fixed the problem.  I wouldn't mind adding this to the README under the Build section since maybe not every dev wants to go poking around in a shell script to figure out how compile.

**Comment by apoelstra** [comment]
@yancyribbens that will only work for people compiling this crate directly, not anybody depending on it.

It's sad that people need to "poke around in a shell script to figure out how to compile", yes, but that's a problem with cargo, not with us.

**Comment by tcharding** [comment]
> I guess the correct answer to "What do I run on MSRV" is look at the cargo-recent.toml to find the latest dependencies that we support.

I get a bit mixed up with lock files, cargo, and pinning, but as far as I can tell we have holes in the test scripts at the moment.

For recent deps in `bitcoin` crate we do:
1. `cp Cargo-recent.lock Cargo.lock`
2. We use `cargo update` to pin some dependencies (whether this changes the lock file being used is not discernible from the shell scripts)
3. We do all the testing with `--locked` (using the current, possibly modified, lockfile)
4. We run `cargo update` and then warn if there are changes to the lock file against the `Cargo-recent.lock`

So, we have two cases that will both pass CI:

1. The pinning was done before the original `Cargo-recent.lock` was made, the pinning commands in the CI script do not change the lock file
2. The pinning was not done before the original `Cargo-recent.lock` file, the pinning commands in the CI script modify the recent lock file

Conclusions:
- In both cases the warnings generated in (4) are invalid for MSRV
- Everytime we need to pin a new dependency or change the pinned version we either have to fix the `Cargo-recent.lock` OR add a pinning command to the CI script - so the CI pinning is **not** the single source of truth. Ouch.

**Comment by apoelstra** [comment]
> * Everytime we need to pin a new dependency or change the pinned version we either have to fix the `Cargo-recent.lock`

`Cargo-recent.lock` should always work, and by definition, will not be affected by new broken dependencies. And our choices of when to update the lockfile shouldn't have anything to do with pinning requirements.

But you are correct that our current mix of pinning and fixed lockfiles is incoherent. We should have CI jobs that use the fixed lockfiles, with no pinning, **and** a CI job that uses the actual latest dependencies along with pinning to make it work.

**Comment by sanket1729** [comment]
Agreed with @apoelstra. 

I wanted something like
```diff
-DEPS="recent minimal"
+DEPS="recent minimal latest"
 
 for dep in $DEPS
 do
-       cp "Cargo-$dep.lock" Cargo.lock
+       if [ "$dep" = latest ];
+       then
+               cargo update
+       else
+               cp "Cargo-$dep.lock" Cargo.lock
+       fi
```

But I am not thrilled that this will increase our CI times by 50%. Moves them from 18 mins to 27 mins. If we are doing this as a part of CI, I fail to see the value of `Cargo-recent.toml` except at the time of release. 

Or maybe another simple lightweight job that just tests --all-features.

**Comment by yancyribbens** [comment]
Maybe the strategy could be outlined and documented some place in a developer section.  IE how to build MSRV and process by which the toml files are updated.

**Comment by yancyribbens** [comment]
>  If we are doing this as a part of CI, I fail to see the value of Cargo-recent.toml except at the time of release.

I'm curious about this as well.

**Comment by apoelstra** [comment]
> I fail to see the value of `Cargo-recent.toml` except at the time of release.

The point is that it's the "latest set of dependencies we support". (And also the latest set of dependencies that I test with locally.) Where "support" has nothing to do with CI, it means the dependency versions we've looked at and are okay with.

We can drop it for now, but eventually the idea is that we'll be using crev and actually keeping this somewhat up to date.

**Comment by yancyribbens** [comment]
closed by #1977
